File: market-domain/src/main/java/com/example/market/domain/catalog/Product.java
package com.example.market.domain.catalog;

import java.time.Instant;
import java.util.Objects;

/**
 * 상품 마스터. 한정판 신발/시계/의류의 *모델* 정보.
 *
 * <p>예: Product = "Air Jordan 1 Retro High OG Chicago" (브랜드 = Nike, 카테고리 = SNEAKERS)<br>
 * Sku = (이 상품의) 270mm / Black</p>
 *
 * <p>Product 자체는 거래 단위가 아니다. 거래는 Sku 단위로 이뤄짐.</p>
 */
public class Product {

    private final ProductId id;
    private final String brand;
    private final String modelName;
    private final String styleCode;          // 예: 555088-101
    private final ProductCategory category;
    private final Instant releaseDate;       // 한정판은 출시일이 의미 있음
    private final String imageUrl;
    private final Instant createdAt;

    private Product(ProductId id, String brand, String modelName, String styleCode,
                    ProductCategory category, Instant releaseDate, String imageUrl,
                    Instant createdAt) {
        this.id = id;
        this.brand = brand;
        this.modelName = modelName;
        this.styleCode = styleCode;
        this.category = category;
        this.releaseDate = releaseDate;
        this.imageUrl = imageUrl;
        this.createdAt = createdAt;
    }

    public static Product create(String brand, String modelName, String styleCode,
                                 ProductCategory category, Instant releaseDate,
                                 String imageUrl, Instant now) {
        Objects.requireNonNull(brand);
        if (brand.isBlank()) throw new IllegalArgumentException("brand must not be blank");
        Objects.requireNonNull(modelName);
        if (modelName.isBlank()) throw new IllegalArgumentException("modelName must not be blank");
        Objects.requireNonNull(category);
        return new Product(ProductId.newId(), brand, modelName, styleCode, category,
                releaseDate, imageUrl, now);
    }

    public static Product restore(ProductId id, String brand, String modelName, String styleCode,
                                  ProductCategory category, Instant releaseDate,
                                  String imageUrl, Instant createdAt) {
        return new Product(id, brand, modelName, styleCode, category, releaseDate, imageUrl, createdAt);
    }

    public ProductId id() { return id; }
    public String brand() { return brand; }
    public String modelName() { return modelName; }
    public String styleCode() { return styleCode; }
    public ProductCategory category() { return category; }
    public Instant releaseDate() { return releaseDate; }
    public String imageUrl() { return imageUrl; }
    public Instant createdAt() { return createdAt; }
}

File: market-domain/src/main/kotlin/com/example/market/domain/catalog/Product.kt
package com.example.market.domain.catalog

import java.time.Instant

/**
 * 상품 마스터. 한정판 신발/시계/의류의 *모델* 정보.
 *
 * 예: Product = "Air Jordan 1 Retro High OG Chicago" (브랜드 = Nike, 카테고리 = SNEAKERS)
 * Sku = (이 상품의) 270mm / Black
 *
 * Product 자체는 거래 단위가 아니다. 거래는 Sku 단위로 이뤄짐.
 *
 * Kotlin `@JvmRecord` 로 컴파일 — Java record 와 동일한 component accessor (`id()`,
 * `brand()`, `modelName()` 등) 을 노출해 호출자 호환성 (Java + Kotlin) 보존. 기존
 * Java class 가 final field + record-style accessor 패턴이라 사실상 record 의 의미와
 * 동일했고, Kotlin 변환으로 record 본연의 형태로 표현.
 */
@JvmRecord
data class Product(
    val id: ProductId,
    val brand: String,
    val modelName: String,
    /** 예: 555088-101. */
    val styleCode: String?,
    val category: ProductCategory,
    /** 한정판은 출시일이 의미 있음. nullable. */
    val releaseDate: Instant?,
    val imageUrl: String?,
    val createdAt: Instant,
) {

    companion object {
        @JvmStatic
        fun create(
            brand: String,
            modelName: String,
            styleCode: String?,
            category: ProductCategory,
            releaseDate: Instant?,
            imageUrl: String?,
            now: Instant,
        ): Product {
            require(brand.isNotBlank()) { "brand must not be blank" }
            require(modelName.isNotBlank()) { "modelName must not be blank" }
            return Product(
                id = ProductId.newId(),
                brand = brand,
                modelName = modelName,
                styleCode = styleCode,
                category = category,
                releaseDate = releaseDate,
                imageUrl = imageUrl,
                createdAt = now,
            )
        }

        @JvmStatic
        fun restore(
            id: ProductId,
            brand: String,
            modelName: String,
            styleCode: String?,
            category: ProductCategory,
            releaseDate: Instant?,
            imageUrl: String?,
            createdAt: Instant,
        ): Product = Product(
            id = id,
            brand = brand,
            modelName = modelName,
            styleCode = styleCode,
            category = category,
            releaseDate = releaseDate,
            imageUrl = imageUrl,
            createdAt = createdAt,
        )
    }
}
